app/models/manageiq/providers/openstack/infra_manager.rb

```diff
@@ -83,4 +83,31 @@ def supports_authentication?(authtype)
   def self.event_monitor_class
     ManageIQ::Providers::Openstack::InfraManager::EventCatcher
   end
+
+  def verify_credentials(auth_type = nil, options = {})
+    auth_type ||= 'default'
+
+    raise MiqException::MiqHostError, "No credentials defined" if missing_credentials?(auth_type)
+
+    options[:auth_type] = auth_type
+    case auth_type.to_s
+    when 'default' then     verify_api_credentials(options)
+    when 'amqp' then        verify_amqp_credentials(options)
+    when 'ssh_keypair' then verify_ssh_keypair_credentials(options)
+    else;                   raise "Invalid OpenStack Authentication Type: #{auth_type.inspect}"
+    end
+  end
+
+  def required_credential_fields(type)
+    case type.to_s
+    when 'ssh_keypair' then [:userid, :auth_key]
+    else                    [:userid, :password]
+    end
+  end
+
+  def verify_ssh_keypair_credentials(_options)
+    hosts.sort_by(&:ems_cluster_id).slice_when { |i, j| i.ems_cluster_id != j.ems_cluster_id }.map { |c| c.find { |h| h.power_state == 'on' } }.all? { |h| h.verify_credentials('ssh_keypair') }
```

Comment:
Can you explain why the hosts have to be grouped by their cluster in order to check if they can validate the ssh keypair credentials?

If you don't have to group by cluster, you could do something like:

```ruby
hosts.where(:power_state => "on").all? { |h| h.verify_credentials("ssh_keypair") }
```

Also, it looks like hosts can throw a `"No credentials defined"` error, so you'd need to `rescue` that.  It might look something more like:

```ruby
hosts.where(:power_state => "on").all? { |h| h.verify_credentials("ssh_keypair") rescue false }
```

... assuming you want all hosts that are "on" to be able to validate ssh keypair.